go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -358,3 +324,232 @@ func updateCondition(conditions []metav1.Condition, cond *metav1.Condition) ([]m
+				}
+			}
+
+			if len(errs) > 0 {
+				return nil, errors.Join(errs...)
+			}
+
+			var err error
+			controllerutil.RemoveFinalizer(cudn, template.FinalizerUserDefinedNetwork)
+			cudn, err = c.udnClient.K8sV1().ClusterUserDefinedNetworks().Update(context.Background(), cudn, metav1.UpdateOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("failed to remove finalizer to ClusterUserDefinedNetwork: %w", err)
+			}
+			klog.Infof("Finalizer removed from ClusterUserDefinedNetwork [%s/%s]", cudn.Namespace, cudn.Name)
+			delete(c.namespaceTracker, cudn.Name)
+		}
+
+		return nil, nil
+	}
+
+	if finalizerAdded := controllerutil.AddFinalizer(cudn, template.FinalizerUserDefinedNetwork); finalizerAdded {
+		udn, err := c.udnClient.K8sV1().ClusterUserDefinedNetworks().Update(context.Background(), cudn, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to add finalizer to ClusterUserDefinedNetwork: %w", err)
+		}
+		klog.Infof("Added Finalizer to ClusterUserDefinedNetwork [%s/%s]", udn.Namespace, udn.Name)
+	}
+
+	selectedNamespaces, err := c.getSelectedNamespaces(cudn.Spec.NamespaceSelector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get selected namespaces: %w", err)
+	}
+
+	var errs []error
+	for nsToDelete := range affectedNamespaces.Difference(selectedNamespaces) {
+		if err := c.deleteNAD(cudn, nsToDelete); err != nil {
+			errs = append(errs, fmt.Errorf("failed to delete NetworkAttachmentDefinition: %w", err))
+		} else {
+			c.namespaceTracker[cudn.Name].Delete(nsToDelete)
+		}
+	}
+
+	var nads []netv1.NetworkAttachmentDefinition
+	for nsToUpdate := range selectedNamespaces {
+		nad, err := c.updateNAD(cudn, nsToUpdate)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to update NetworkAttachmentDefinition: %w", err))
+		} else {
+			if _, exist := c.namespaceTracker[cudn.Name]; !exist {
+				c.namespaceTracker[cudn.Name] = sets.New[string]()
+			}
+			c.namespaceTracker[cudn.Name].Insert(nsToUpdate)
+			nads = append(nads, *nad)
+		}
+	}
+
+	return nads, errors.Join(errs...)
+}
+
+// getSelectedNamespaces list all selected namespaces according to given selector and create
+// a set of the selected namespaces keys.
+func (c *Controller) getSelectedNamespaces(sel metav1.LabelSelector) (sets.Set[string], error) {
+	selectedNamespaces := sets.Set[string]{}
+	labelSelector, err := metav1.LabelSelectorAsSelector(&sel)
+	if err != nil {
+		return nil, fmt.Errorf("faild to create label-selector: %w", err)
+	}
+	selectedNamespacesList, err := c.namespaceInformer.Lister().List(labelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("faild to list namespaces: %w", err)
+	}
+	for _, selectedNs := range selectedNamespacesList {
+		selectedNamespaces.Insert(selectedNs.Name)
+	}
+	return selectedNamespaces, nil
+}
+
+func (c *Controller) updateClusterUDNStatus(cudn *userdefinednetworkv1.ClusterUserDefinedNetwork, nads []netv1.NetworkAttachmentDefinition, syncError error) error {
```

Comment:
with cUDN, I have got an idea:
we could just create a special error type for "NetworkAttachmentDefinitionDeleted" reason, and check `nad.DeletionTimestamp` inside `syncUserDefinedNetwork` function instead of doing that in the `updateStatus` function. Then we won't need to pass nads around, but just use errors list, wdyt?

Reply:
added to follow-up list